plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/flex/FlexDeployPreferences.java

```diff
@@ -17,27 +17,36 @@
 package com.google.cloud.tools.eclipse.appengine.deploy.flex;
 
 import com.google.cloud.tools.eclipse.appengine.deploy.DeployPreferences;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Strings;
 import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ProjectScope;
+import org.eclipse.core.runtime.preferences.IEclipsePreferences;
 import org.osgi.service.prefs.BackingStoreException;
 
 public class FlexDeployPreferences extends DeployPreferences {
```

Comment:
The hierarchy of preferences is confusing me. Does this match how Eclipse stores the preferences? E.g. is there one deploy preferences that might or might not be a flex deploy preferences, or are there multiple deploy preferences of different types?

Reply:
It is a different hierarchy than the project-workspace-instace preferences (where you try to find something in a project, moves on to a workspace if it is not there, and so on). First, we have a common set of deploy parameters (account email, GCP project, etc) as `DeployPreferences` that can be applied to both standard and flex. (Incidentally, this common set is all that the standard needs, so we have no `StandardDeployPreferences`.) Flex needs one additional param, the app.yaml path, so it extends the common set. The global flex preferences is for the arbitrary WAR/JAR deployment, which additionally needs a WAR/JAR path, so it extends the flex preferences. (However, it stores preferences in the global scope, as it is not bound to any project.) In some sense, they satisfy the IS-A relationship. (`FlexGlobalDeployPreferences` is a `FlexDeployPreferences`, and so on.)

If you come up with a better idea, let me just know.